File: python/src/trees/invert_binary_tree.py
from tree import Tree


def invert_tree(root: Tree) -> Tree:
    if not root:
        return None
    stack = [root]
    while stack:
        node = stack.pop()
        node.left, node.right = node.right, node.left
        if node.left:
            stack.append(node.left)
        if node.right:
            stack.append(node.right)
    return root


def invert_tree_recursive(root: Tree) -> Tree:
    if not root:
        return None
    root.left, root.right = (
        invert_tree_recursive(root.right),
        invert_tree_recursive(root.left),
    )
    return root

File: typescript/src/trees/invert-binary-tree.ts
import { Tree } from "./tree";

export function invertTree(root: Tree): Tree {
  if (!root) {
    return null;
  }
  const stack = [root];
  while (stack.length > 0) {
    const node = stack.pop()!;
    [node.left, node.right] = [node.right, node.left];
    if (node.left) {
      stack.push(node.left);
    }
    if (node.right) {
      stack.push(node.right);
    }
  }
  return root;
}

export function invertTreeRecursive(root: Tree): Tree {
  if (!root) {
    return null;
  }
  [root.left, root.right] = [
    invertTreeRecursive(root.right),
    invertTreeRecursive(root.left),
  ];
  return root;
}
